class_name GatewayManagerClient
extends BaseMultiplayerEndpoint


signal account_creation_result_received(user_id: int, result_code: int, data: Dictionary)
signal login_succeeded(account_info: Dictionary, _worlds_info: Dictionary)
signal response_received(response: Dictionary)

var worlds_info: Dictionary


func _ready() -> void:
	var configuration: Dictionary = ConfigFileUtils.load_section(
		"gateway-manager-client",
		CmdlineUtils.get_parsed_args().get("config", "res://data/config/gateway_config.cfg")
	)

	var tls_options: TLSOptions = null
	if configuration.has("certificate_path") and not configuration.certificate_path.is_empty():
		var cert = load(configuration.certificate_path)
		if cert:
			tls_options = TLSOptions.client(cert)

	create(Role.CLIENT, configuration.address, configuration.port, tls_options)


func _connect_multiplayer_api_signals(api: SceneMultiplayer) -> void:
	api.connected_to_server.connect(_on_connection_succeeded)
	api.connection_failed.connect(_on_connection_failed)
	api.server_disconnected.connect(_on_server_disconnected)


func _on_connection_succeeded() -> void:
	print("Successfully connected to the Gateway Manager as %d!" % multiplayer.get_unique_id())


func _on_connection_failed() -> void:
	print("Failed to connect to the Gateway Manager as Gateway.")
	# Try to reconnect.
	get_tree().create_timer(15.0).timeout.connect(_ready)


func _on_server_disconnected() -> void:
	print("Gateway Manager disconnected.")
	# Try to reconnect.
	get_tree().create_timer(15.0).timeout.connect(_ready)


@rpc("authority")
func update_worlds_info(_worlds_info: Dictionary) -> void:
	worlds_info = _worlds_info
	# Sanitizar puertos (evitar 0) por si algún world no envió correctamente.
	for w_id in worlds_info.keys():
		var w: Dictionary = worlds_info[w_id]
		if not (w is Dictionary):
			continue
		var p: int = int(w.get("port", 0))
		if p <= 0:
			w["port"] = 8087
		worlds_info[w_id] = w


@rpc("authority")
func fetch_auth_token(target_peer: int, auth_token: String, _address: String, _port: int) -> void:
	response_received.emit(
		{"t-id": target_peer, "token": auth_token, "address": _address, "port": _port}
	)
	#gateway.connected_peers[target_peer]["token_received"] = true
	#gateway.fetch_auth_token.rpc_id(target_peer, auth_token, _address, _port)


@rpc("any_peer")
func login_request(_peer_id: int, _username: String, _password: String) -> void:
	pass


@rpc("authority")
func login_result(peer_id: int, result: Dictionary) -> void:
	if result.has("error"):
		response_received.emit(
			{"t-id": peer_id, "a": result, "w": worlds_info, "error": result}
		)
		#gateway.login_result.rpc_id(peer_id, result["error"])
	else:
		#gateway.login_result.rpc_id(peer_id, 0)
		#login_succeeded.emit(peer_id, result, worlds_info)
		response_received.emit(
			{"t-id": peer_id, "a": result, "w": worlds_info}
		)


@rpc("any_peer")
func create_account_request(_peer_id: int, _username: String, _password: String, _is_guest: bool) -> void:
	pass


@rpc("authority")
func account_creation_result(peer_id: int, result_code: int, result: Dictionary) -> void:
	if result_code == OK:
		#login_succeeded.emit(peer_id, result, worlds_info)
		response_received.emit(
			{"t-id": peer_id, "a": result, "w": worlds_info}
		)
	#gateway.account_creation_result.rpc_id(peer_id, result_code)


@rpc("any_peer")
func create_player_character_request(_peer_id: int , _username: String, _character_data: Dictionary, _world_id: int) -> void:
	pass


@rpc("authority")
func player_character_creation_result(peer_id: int, result: Dictionary) -> void:
	response_received.emit(
		{"t-id": peer_id, "data": result}
	)
	#gateway.player_character_creation_result.rpc_id(
		#peer_id, result_code
	#)


@rpc("any_peer")
func request_player_characters(_peer_id: int, _username: String, _world_id: int) -> void:
	pass


@rpc("authority")
func receive_player_characters(peer_id: int, player_characters: Dictionary) -> void:
	response_received.emit(
		{"t-id": peer_id, "data": player_characters}
	)
	#gateway.receive_player_characters.rpc_id(peer_id, player_characters)


@rpc("any_peer")
func request_login(_peer_id: int, _username: String, _world_id: int, _character_id: int) -> void:
	pass


@rpc("any_peer")
func peer_disconnected_without_joining_world(_account_name: String) -> void:
	pass


#@rpc("any_peer")
#func request_world_info() -> void:
	#pass
#
#
#@rpc("authority")
#func receive_world_info(world_info: Dictionary) -> void:
	#response_received.emit(worlds_info)
